feat: sqlx sqlite expose de/serialize

mattrighetti opened the pull request:
This PR exposes two new functions for sqlite (#192):
- `serialize`
- `deserialize`

I've tried to follow what @abonander suggested https://github.com/launchbadge/sqlx/issues/192#issuecomment-2652517185

### Related docs
- [serialize](https://www.sqlite.org/c3ref/serialize.html)
- [deserialize](https://www.sqlite.org/c3ref/deserialize.html)
- [sqlite malloc](https://www.sqlite.org/c3ref/free.html)

### Serialize

`serialize_nocopy` uses `sqlite3_serialize` with the flag `SQLITE_SERIALIZE_NOCOPY` enabled so no new buffer is allocated. It is implemented on `LockedHandle` so that we're sure the buffer does not outlive the handle lifetime as noted in the sqlite docs:

> After the call, if the SQLITE_SERIALIZE_NOCOPY bit had been set, the returned buffer content will remain accessible and unchanged until either the next write operation on the connection or when the connection is closed, and applications must not modify the buffer. If the bit had been clear, the returned buffer will not be accessed by SQLite after the call.

I've then implemented the `serialize` function on `SqliteConnection` that basically locks the handle and copies the buffer returned by `serialize_nocopy` into an owned type `Vec<u8>` that will be returned to the caller.

### Deserialize

`deserialize` requires a bit more work. First of all I've created a wrapper type `SqliteOwnedBuf` for a buffer allocated with `sqlite3_malloc`. That buffer is useful in `sqlite3_deserialize` since it enables us to use the `SQLITE_DESERIALISE_FREEONCLOSE` flag that will notify `sqlite3_deserialize` to manage the buffer and free the allocation when it's done with it.

### Notes

This still needs testing, but I wanted to first understand if there are other changes needed for this implementation.

Also, I really doubt that the `Err` I've implemented are correct, I'd need some guidance on how we should handle those failures

Looking forward to your feedback, thanks!

abonander commented on sqlx-sqlite/src/connection/mod.rs:
```diff
@@ -201,6 +204,90 @@ impl SqliteConnection {
+    pub async fn serialize(&mut self, schema: &str) -> Result<Option<Vec<u8>>, Error> {
+        let mut locked_handle = self.lock_handle().await?;
+
+        Ok(locked_handle.serialize_nocopy(schema)?.map(Vec::from))
```
> I've then implemented the `serialize` function on SqliteConnection that basically locks the handle and copies the buffer returned by `serialize_nocopy` into an owned type Vec<u8> that will be returned to the caller.

That's not how it should work. The problem with `SQLITE_SERIALIZE_NOCOPY` is it only works if the database came _from_ `sqlite3_deserialize()` initially, making it impossible to get a serialization of the database if one doesn't already exist.

It can be used as an _optimization_, though. If `serialize_nocopy()` returns a valid pointer, that can just be used. However, we can leave this as a choice up to the user. If they want an owned buffer anyway, we might as well let SQLite allocate and fill it, and just return that.

mattrighetti replied:
So if a user want to use the `NOCOPY` version it has to call if from the locked handle, right?

The one implemented on the `SqliteConnection` is going to allocate a buffer using `sqlite_malloc`.

Feels a bit odd to have those two different versions on different `struct`. We could give the user a boolean flag to chose one between the two directly on `SqliteConnection` but that will still require me to get ownership of the `NOCOPY` buffer because that lifetime is tied to the `LockedHandle`.

abonander replied:
You _can't_ get ownership of the `NOCOPY` buffer. SQLite maintains ownership of it and will continue using it. It is truly a just borrow. And copying it to another buffer defeats the purpose of that.

The other thing that I specified in #192 is that the `serialize()` and `deserialize()` calls should happen on the worker thread, because they can potentially involve non-trivial blocking work that shouldn't happen in an `async fn` or it'll block the executor.

Moreso for `serialize()` because it may have to wait for concurrent write transactions from other connections to complete, and likely read from the database file on-disk. `deserialize()` _seems_ safe to execute directly, 

I'd honestly be more comfortable just omitting `NOCOPY` from this first pass of the API, because it may have more caveats than I initially thought. Reading through the implementation, it also waits for concurrent transactions to complete _in the `NULL` return case_, because it executes `PRAGMA <schema>.page_count` to get the size of the database to return in the `sqlite3_int64 *piSize` out-pointer: https://github.com/sqlite/sqlite/blob/eb3a069fc82e53a40ea63076d66ab113a3b2b0c6/src/memdb.c#L793

I actually confirmed this by writing a short little C program. It does this regardless of the `NO_COPY` flag; in the regular copying case, it uses the return value of `PRAGMA page_count` both to know the size of the buffer to allocate, and the number of pages to enumerate. And executing the `PRAGMA` statement opens a read transaction.

abonander commented on sqlx-sqlite/src/connection/mod.rs:
```diff
@@ -201,6 +204,90 @@ impl SqliteConnection {
+    pub async fn deserialize(
+        &mut self,
+        schema: &str,
+        data: Vec<u8>,
```
This doesn't need to be an owned `Vec<u8>` if we're copying it to a buffer anyway.

abonander replied:
I had an idea for a trait that would be implemented for `&[u8]` and `Vec<u8>` and `SqliteOwnedBuf` (possibly others as well) that would decide whether to copy to another buffer or not. But that could be part of a later iteration of this API.

abonander commented on sqlx-sqlite/src/connection/mod.rs:
```diff
@@ -201,6 +204,90 @@ impl SqliteConnection {
+    ) -> Result<(), Error> {
+        let mut locked_handle = self.lock_handle().await?;
+
+        let c_schema = CString::new(schema).expect("invalid cstring contains nul byte");
```
This should not panic. It's possible that this may be exposed to the user somehow, and the application shouldn't crash if they pass an invalid string. I would return [`Error::Configuration`](https://docs.rs/sqlx/latest/sqlx/enum.Error.html#variant.Configuration) or a new variant.

mattrighetti replied:
For now I've created an ad-hoc error for the sqlite buffer and I'm returning that boxed in `Error::Configuration`

abonander commented on sqlx-sqlite/src/connection/mod.rs:
```diff
@@ -201,6 +204,90 @@ impl SqliteConnection {
+            flags |= SQLITE_DESERIALIZE_RESIZEABLE;
+        }
+
+        let sqlite_buf = SqliteOwnedBuf::try_from(data).expect("could not initialize buffer");
```
Same here re. panicking. For one, this will fail if the length of `data` is zero, because `sqlite3_malloc()` returns a null pointer on a zero allocation size.

mattrighetti commented:
Update: I've moved `serialize()` and `deserialize()` on the worker thread but I have little knowledge of the codebase so feel free to point me in the right direction.

I had to introduce two new `Command` enumerations `Serialize` + `Deserialize` so that they can be used and passed in the worker `tx` channel.

`serialize()` probably requires more work, as @abonander pointed out

>  it may have to wait for concurrent write transactions from other connections to complete

plus, I've implemented `Send + Sync for SqliteOwnedBuf` just to silence the compiler complaints because `NonNull<u8>` is not thread-safe ofc.

abonander commented on sqlx-sqlite/src/connection/serialize.rs:
```diff
@@ -0,0 +1,33 @@
+
+/// Serializes the SQLite database into a byte vector.
+///
+/// This function returns the serialized bytes of the database for the specified `schema`
+/// only if `deserialize` was previously called. If no data is available, it returns `None`.
```
```suggestion
```

Outdated remark.

abonander commented:
@mattrighetti on further review, there were a lot of things I wanted to change and it would have been unwieldy to do it all in review comments. I took the liberty of pushing a commit with the changes:

* expanded docs better matching the style of the rest of the project
* better error handling
* integration tests

Take a look and let me know what you think.

mattrighetti commented:
lgtm!

Not 100% convinced about the `serialize(Option<&str>)` but I guess it makes sense for this first iteration.

In the future we could introduce a new function `serialize_schema(&str)` where we explicitly ask for a schema while keeping `serialize()` as "main" schema serialization.
